File: src/main/java/com/lessmarkup/engine/recordmodel/RecordModelDefinitionImpl.java
package com.lessmarkup.engine.recordmodel;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.engine.scripting.ScriptHelper;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.helpers.PropertyDescriptor;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.helpers.TypeHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.exceptions.RecordValidationException;
import com.lessmarkup.interfaces.recordmodel.*;
import com.lessmarkup.interfaces.system.EngineConfiguration;

import java.util.*;
import net.tanesha.recaptcha.ReCaptcha;
import net.tanesha.recaptcha.ReCaptchaFactory;
import net.tanesha.recaptcha.ReCaptchaResponse;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

@Component
@Scope("prototype")
public class RecordModelDefinitionImpl implements RecordModelDefinition {

    private String titleTextId;
    private String moduleType;
    private Class<? extends RecordModel> modelType;
    private Class<? extends DataObject> dataType;
    private boolean submitWithCaptcha;
    private final List<InputFieldDefinition> fields = new ArrayList<>();
    private final List<RecordColumnDefinition> columns = new ArrayList<>();
    private String id;
    private RecordModel recordModelInstance;
    
    @SuppressWarnings("unchecked")
    public void initialize(Class<? extends RecordModel> modelType, String moduleType) {
        
        this.recordModelInstance = DependencyResolver.resolve(modelType);
        
        this.titleTextId = this.recordModelInstance.getTitleTextId();
        this.moduleType = moduleType;
        this.modelType = modelType;
        this.dataType = this.recordModelInstance.getDataType();
        this.submitWithCaptcha = this.recordModelInstance.getSubmitWithCaptcha();

        for (PropertyDescriptor property : TypeHelper.getProperties(modelType)) {
            InputField inputField = property.getAnnotation(InputField.class);

            if (inputField != null) {
                InputFieldDefinition definition = new InputFieldDefinition(property, inputField);
                this.fields.add(definition);
            }
            
            RecordColumn recordColumn = property.getAnnotation(RecordColumn.class);

            if (recordColumn != null) {
                RecordColumnDefinition definition = new RecordColumnDefinition();
                definition.initialize(recordColumn, property);
                this.columns.add(definition);
                
                if (recordColumn.cellTemplate() != null) {
                    definition.setCellTemplate(recordColumn.cellTemplate());
                }
            }
        }
    }
    
    @Override
    public String getTitleTextId() {
        return this.titleTextId;
    }

    @Override
    public String getModuleType() {
        return this.moduleType;
    }

    @Override
    public Class<? extends RecordModel> getModelType() {
        return this.modelType;
    }

    @Override
    public Class<? extends DataObject> getDataType() {
        return this.dataType;
    }

    @Override
    public String getId() {
        return this.id;
    }
    
    void setId(String id) {
        this.id = id;
    }

    @Override
    public List<InputFieldDefinition> getFields() {
        return this.fields;
    }

    @Override
    public List<RecordColumnDefinition> getColumns() {
        return this.columns;
    }

    private static final String ChallengeFieldKey = "-RecaptchaChallenge-";
    private static final String ResponseFieldKey = "-RecaptchaResponse-";

    @Override
    public void validateInput(JsonElement objectToValidate, boolean isNew) throws RecordValidationException {
        
        if (objectToValidate == null || objectToValidate.isJsonNull()) {
            throw new IllegalArgumentException();
        }
        
        if (this.submitWithCaptcha) {
            if (!objectToValidate.isJsonObject()) {
                throw new RecordValidationException("Cannot validate captcha");
            }
            
            JsonObject propertiesObject = objectToValidate.getAsJsonObject();

            String challengeValue = propertiesObject.getAsJsonPrimitive(ChallengeFieldKey).toString();
            String responseValue = propertiesObject.getAsJsonPrimitive(ResponseFieldKey).toString();

            EngineConfiguration engineConfiguration = RequestContextHolder.getContext().getEngineConfiguration();
            ReCaptcha reCaptcha = ReCaptchaFactory.newReCaptcha(engineConfiguration.getRecaptchaPublicKey(), engineConfiguration.getRecaptchaPrivateKey(), false);
            ReCaptchaResponse response = reCaptcha.checkAnswer(RequestContextHolder.getContext().getRemoteAddress(), challengeValue, responseValue);
            
            if (!response.isValid()) {
                throw new RecordValidationException(response.getErrorMessage());
            }
        }
        
        for (InputFieldDefinition field : this.fields) {
            if (!field.isRequired()) {
                continue;
            }
            
            if (field.getType() == InputFieldType.FILE && !isNew) {
                continue;
            }
            
            if (!StringHelper.isNullOrEmpty(field.getVisibleCondition()) && !ScriptHelper.evaluateExpression(field.getVisibleCondition(), objectToValidate)) {
                continue;
            }
            
            if (!StringHelper.isNullOrEmpty(field.getReadOnlyCondition()) && ScriptHelper.evaluateExpression(field.getReadOnlyCondition(), objectToValidate)) {
                continue;
            }
            
            if (!objectToValidate.isJsonObject()) {
                continue;
            }
            
            JsonElement fieldValue = objectToValidate.getAsJsonObject().get(StringHelper.toJsonCase(field.getProperty().getName()));
            
            if (fieldValue != null && fieldValue.toString().length() > 0) {
                continue;
            }
            
            String errorText = LanguageHelper.getText(Constants.ModuleType.MAIN, TextIds.PROPERTY_MUST_BE_SPECIFIED);
            String fieldText = field.getTextId() == null ? "" : LanguageHelper.getText(moduleType, field.getTextId());
            
            throw new RecordValidationException(String.format(errorText, fieldText));
        }
    }

    @Override
    public boolean isSubmitWithCaptcha() {
        return this.submitWithCaptcha;
    }

    @Override
    public ModelCollection createModelCollection() {
        return this.recordModelInstance.createCollection();
    }
}

File: src/main/scala/com/lessmarkup/engine/recordmodel/RecordModelDefinitionImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.recordmodel

import java.security.MessageDigest
import java.util.Base64

import com.google.gson.{JsonElement, JsonObject}
import com.lessmarkup.engine.scripting.ScriptHelper
import com.lessmarkup.framework.helpers.{DependencyResolver, LanguageHelper, StringHelper, TypeHelper}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.{InputField, InputFieldType, RecordColumn}
import com.lessmarkup.interfaces.data.DataObject
import com.lessmarkup.interfaces.exceptions.RecordValidationException
import com.lessmarkup.interfaces.recordmodel._
import com.lessmarkup.interfaces.system.EngineConfiguration
import com.lessmarkup.{Constants, TextIds}
import net.tanesha.recaptcha.{ReCaptcha, ReCaptchaFactory, ReCaptchaResponse}

object RecordModelDefinitionImpl {
  private val ChallengeFieldKey: String = "-RecaptchaChallenge-"
  private val ResponseFieldKey: String = "-RecaptchaResponse-"
}

class RecordModelDefinitionImpl(modelType: Class[_ <: RecordModel[_]], moduleType: String, index: Int) extends RecordModelDefinition {

  private val recordModelInstance = DependencyResolver(modelType)
  private val dataType: Class[_ <: DataObject] = recordModelInstance.getDataType
  private val id = createDefinitionId
  private val properties = TypeHelper.getProperties(modelType)
  private val fields = properties
    .map(p => (p, p.getAnnotation(classOf[InputField])))
    .filter(_._2 != null)
    .map { case (p, f) => new InputFieldDefinition(p, f)}
    .toList
  private val columns = properties
    .map(p => (p, p.getAnnotation(classOf[RecordColumn])))
    .filter(_._2 != null)
    .map { case (p, c) => new RecordColumnDefinition(c, p) }
    .toList

  private def createDefinitionId: String = {
    val messageDigest = MessageDigest.getInstance("SHA-256")
    val idString: String = getModelType.getName + index
    val bytes: Array[Byte] = messageDigest.digest(idString.getBytes)
    Base64.getEncoder.encodeToString(bytes)
  }

  def getTitleTextId: String = recordModelInstance.getTitleTextId

  def getModuleType: String = moduleType

  def getModelType: Class[_ <: RecordModel[_]] = modelType

  def getDataType: Class[_ <: DataObject] = dataType

  def getId: String = id

  def getFields = fields

  def getColumns = columns

  def validateInput(objectToValidate: JsonElement, isNew: Boolean) {
    if (objectToValidate == null || objectToValidate.isJsonNull) {
      throw new IllegalArgumentException
    }
    if (isSubmitWithCaptcha) {
      if (!objectToValidate.isJsonObject) {
        throw new RecordValidationException("Cannot validate captcha")
      }
      val propertiesObject: JsonObject = objectToValidate.getAsJsonObject
      val challengeValue: String = propertiesObject.getAsJsonPrimitive(RecordModelDefinitionImpl.ChallengeFieldKey).toString
      val responseValue: String = propertiesObject.getAsJsonPrimitive(RecordModelDefinitionImpl.ResponseFieldKey).toString
      val engineConfiguration: EngineConfiguration = RequestContextHolder.getContext.getEngineConfiguration
      val reCaptcha: ReCaptcha = ReCaptchaFactory.newReCaptcha(engineConfiguration.getRecaptchaPublicKey, engineConfiguration.getRecaptchaPrivateKey, false)
      val response: ReCaptchaResponse = reCaptcha.checkAnswer(RequestContextHolder.getContext.getRemoteAddress, challengeValue, responseValue)
      if (!response.isValid) {
        throw new RecordValidationException(response.getErrorMessage)
      }
    }

    if (!objectToValidate.isJsonObject) {
      return
    }

    fields
      .filter(f => f.isRequired
        && (f.getType != InputFieldType.FILE || !isNew)
        && (StringHelper.isNullOrEmpty(f.getVisibleCondition) || ScriptHelper.evaluateExpression(f.getVisibleCondition, objectToValidate))
        && (StringHelper.isNullOrEmpty(f.getReadOnlyCondition) || !ScriptHelper.evaluateExpression(f.getReadOnlyCondition, objectToValidate))
        && {
          val fieldValue: JsonElement = objectToValidate.getAsJsonObject.get(StringHelper.toJsonCase(f.getProperty.getName))
          fieldValue == null || fieldValue.toString.length == 0
        }
      )
      .foreach(f => {
        val errorText: String = LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.PROPERTY_MUST_BE_SPECIFIED)
        val fieldText: String = if (f.getTextId == null) "" else LanguageHelper.getText(moduleType, f.getTextId)
        throw new RecordValidationException(String.format(errorText, fieldText))
      })
  }

  def isSubmitWithCaptcha: Boolean = recordModelInstance.getSubmitWithCaptcha

  def createModelCollection: ModelCollection[_] = {
    recordModelInstance.createCollection
  }
}